cornice/ext/swagger.py

```diff
@@ -0,0 +1,207 @@
+                        paramType = "path"
+                    else:
+                        paramType = "query"
+                parameter["in"] = paramType
+                parameter["name"] = name
+                if hasattr(attr, "description"):
+                    parameter["description"] = getattr(attr, "description")
+                parameter["required"] = getattr(attr, "required")
+
+                type_ = ""
+                if hasattr(attr, "typ"):
+                    type_ = attr.typ.__class__.__name__
+
+                parameter["type"] = type_
+
+            ret.append(parameter)
+
+    # body
+    swag = cornice.ext.swagger_model.SwaggerModel()
+    entry = swag.to_swagger(schema.colander_schema)
+
+    if swag.models:
+        parameter = dict()
+        parameter["in"] = "body"
+        parameter["name"] = "body"
+        parameter["description"] = schema.__doc__
+        parameter["schema"] = entry
+        ret.append(parameter)
+    return ret, swag.models
+
+
+# At some point, we may have some operation parameters which we"d
+# like to update with a new (only if their names are the same
+# and in the same location)
+def _update_op_param(operation, new):
+    if "name" not in new or "in" not in new:
+        return None
+    for old in operation["parameters"]:
+        if "name" in old and new["name"] == old[
+                "name"] and "in" in old and new["in"] == old["in"]:
+            old.update(new.copy())
+            return True
+    # We've made it through the loop without finding a match, so
+    # add it
+    operation["parameters"].append(new.copy())
+    return False
+
+
+def generate_swagger_spec(services, title, version, **kwargs):
+    """Utility to turn cornice web services into a Swagger-readable file.
+
+    See https://helloreverb.com/developers/swagger for more information.
+    https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md
+    """
+
+    doc = {
+        "swagger": "2.0",
+        "info": {
+            "title": title,
+            "version": str(version)
+        },
+        "paths": {},
+        "tags": [],
+        "definitions": {},
+        "basePath": kwargs.get("base_path", "/")
+    }
+
+    # Handle all the non-required args if passed per spec:
+    # https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#swagger-object-
+    for k, v in list(kwargs.items()):
+        doc[k] = v
+
+    # To control uniqueness of our tags, make a dict
+    tags_dict = dict()
+    definitions = {}
+
+    # sort our services first
+
+    # Loop through all our service endpoint resources
+    for service in services:
+        # Create a tag obejct to add to our base doc later
+        tag_name = service.path.split("/")[1]
+        service_tag = dict(name=tag_name)
+
+        path = {"parameters": []}
+
+        # Get path parameters from looking at the path
+        service_path = service.path
+        parameter = dict()
+
+        # Do we have parameters in the path?
```

Comment:
Lines 117-128 here should be deleted as they

1. duplicate lines 158-168
2. Introduce a bug where the autodetected path parameters in this block are not "updated" in future code (ie Colander SchemaNode), causing duplicate Parameters to show up in Swagger JSON/UI